prov/sockets/src/sock_ep.c

```diff
@@ -879,11 +884,34 @@ static int sock_ep_bind(struct fid *fid, struct fid *bfid, uint64_t flags)
 	return 0;
 }
 
+int sock_ep_setopflags(struct sock_ep *sock_ep, uint64_t flags)
+{
+	if ((flags & FI_TRANSMIT) && (flags & FI_RECV)) {
+		SOCK_LOG_ERROR("Both Tx/Rx flags cannot be specified\n");
+		return -FI_EINVAL;
+	} else if (flags & FI_TRANSMIT) {
+		sock_ep->tx_attr.op_flags = flags;
+		sock_ep->tx_attr.op_flags &= ~FI_TRANSMIT;
+		if (!(flags & (FI_INJECT_COMPLETE | FI_TRANSMIT_COMPLETE |
+		     FI_DELIVERY_COMPLETE)))
+			sock_ep->tx_attr.op_flags |= FI_TRANSMIT_COMPLETE;
+	} else if (flags & FI_RECV) {
+		sock_ep->rx_attr.op_flags = flags;
+		sock_ep->rx_attr.op_flags &= ~FI_RECV;
+	} else {
+		SOCK_LOG_ERROR("Tx/Rx flags not specified\n");
+		return -FI_EINVAL;
+	}
+	return 0;
+}
```

Comment:
Can you look at changing this function to take as input the tx_attr and rx_attr in place of sock_ep?  Then call it from sock_ctx_control.  I'd like to avoid duplicating that complex if statement that adds FI_TRANSMIT_COMPLETE in multiple places.